adam/learner/pursuit.py

```diff
@@ -657,16 +599,655 @@ def compute_hypothesis_id(h: PerceptionGraphPattern) -> str:
+            )
+        else:
+            raise RuntimeError(
+                "Empty template variables:", self.template_variables_to_object_match_nodes
+            )
+
+    def _candidate_perceptions(self, observed_perception_graph) -> List[PerceptionGraph]:
+        # The directions of edges in the perception graph are not necessarily meaningful
+        # from the point-of-view of hypothesis generation, so we need an undirected copy
+        # of the graph.
+        perception_graph = observed_perception_graph.copy_as_digraph()
+        # as_view=True loses s
+        perception_graph_as_undirected = observed_perception_graph.copy_as_digraph().to_undirected(
+            as_view=False
+        )
+        # The core of our hypothesis for the semantics of a preposition is all nodes
+        # along the shortest path between the two objects involved in the perception graph.
+        hypothesis_spine_nodes: ImmutableSet[PerceptionGraphNode] = immutableset(
+            flatten(
+                # if there are multiple paths between the object match nodes,
+                # we aren't sure which are relevant, so we include them all in our hypothesis
+                # and figure we can trim out irrelevant stuff as we make more observations.
+                all_shortest_paths(
+                    perception_graph_as_undirected,
+                    self.object_match_node_for_ground,
+                    self.object_match_node_for_modified,
+                )
+            )
+        )
+
+        # Along the core of our hypothesis we also want to collect the predecessors and successors
+        hypothesis_nodes_mutable = []
+        for node in hypothesis_spine_nodes:
+            if node not in [
+                self.object_match_node_for_ground,
+                self.object_match_node_for_modified,
+            ]:
+                for successor in perception_graph.successors(node):
+                    if not isinstance(successor, ObjectPerception):
+                        hypothesis_nodes_mutable.append(successor)
+                for predecessor in perception_graph.predecessors(node):
+                    if not isinstance(predecessor, ObjectPerception):
+                        hypothesis_nodes_mutable.append(predecessor)
+
+        hypothesis_nodes_mutable.extend(hypothesis_spine_nodes)
+
+        # We wrap the nodes in an immutable set to remove duplicates
+        hypothesis_nodes = immutableset(hypothesis_nodes_mutable)
+
+        preposition_pattern_graph = digraph_with_nodes_sorted_by(
+            networkx_utils.subgraph(
+                observed_perception_graph.copy_as_digraph(), nodes=hypothesis_nodes
+            ),
+            _graph_node_order,
+        )
+        return [PerceptionGraph(preposition_pattern_graph)]
+
+    def _candidate_meanings(
+        self, observed_perception_graph: PerceptionGraph
+    ) -> Sequence[PrepositionPattern]:
+        return [
```

Comment:
This is the same in both versions. If we think it will always be the same, we should hoist it up to the superclass.